openghg/objectstore/metastore/_classic_metastore.py

```diff
@@ -0,0 +1,159 @@
+"""
+This module implements a MetaStore based on TinyDB, with custom
+middleware based on `CachingMiddleware`.
+
+The metastore can be accessed given a bucket and data type.
+ In a context manager, use `open_metastore`. Outside of a context
+ manager, use `ClassicMetaStore.from_bucket`; in this case, the
+metastore must be closed using the `close` method.
+
+Closing the metastore is necessary, since `CachingMiddleware`
+doesn't write to disk unless at least 1000 writes have been made.
+
+The `ClassicMetaStore` is organised in the same way that the
+metastore was organised in OpenGHG <= v 6.2: there are separate
+TinyDB databases for each data type.
+
+Besides taking a data type parameter in the `__init__` method,
+`ClassicMetaStore` only extends `TinyDBMetastore` by adding
+a `from_bucket` class method, and a `close` method.
+These methods are only needed by `store.BaseStore`.
+
+"""
+from __future__ import annotations
+
+from collections.abc import Generator
+from contextlib import contextmanager
+import json
+from typing import Literal, Optional, TypeVar
+
+import tinydb
+from tinydb.middlewares import CachingMiddleware
+
+from openghg.objectstore import exists, get_object, set_object_from_json
+from openghg.objectstore.metastore._metastore import TinyDBMetaStore
+from openghg.types import MetastoreError
+
+
+object_store_data_classes = {
+    "surface": {"_root": "ObsSurface", "_uuid": "da0b8b44-6f85-4d3c-b6a3-3dde34f6dea1"},
+    "column": {"_root": "ObsColumn", "_uuid": "5c567168-0287-11ed-9d0f-e77f5194a415"},
+    "emissions": {"_root": "Emissions", "_uuid": "c5c88168-0498-40ac-9ad3-949e91a30872"},
+    "footprints": {"_root": "Footprints", "_uuid": "62db5bdf-c88d-4e56-97f4-40336d37f18c"},
+    "boundary_conditions": {"_root": "BoundaryConditions", "_uuid": "4e787366-be91-4fc5-ad1b-4adcb213d478"},
+    "eulerian_model": {"_root": "EulerianModel", "_uuid": "63ff2365-3ba2-452a-a53d-110140805d06"},
+}
+
+
+def get_metakey(data_type: str) -> str:
+    """Return the metakey for a given data type."""
+    try:
+        result = object_store_data_classes[data_type]
+    except KeyError:
+        return "default"
+    return f"{result['_root']}/uuid/{result['_uuid']}/metastore"
+
+
+class ObjectStorage(tinydb.Storage):
```

Comment:
I was more thinking that this object is tied specifically to the Metastore at the moment but "ObjectStorage" as a term to me describes the wider concept of the store encompassing the Metastore and the Data itself?

Reply:
Yeah but it is for storing the TinyDB in the object store.

Reply:
Yes I agree on your TinyDB point - I'm more querying the "ObjectStore" name - what does that mean to us when we're defining our system?

Reply:
Maybe `BucketKeyStorage` since its subclassing `Storage` and the main difference from just writing to a particular file is that it's determining where to store things given a bucket and a key.

Reply:
Also, I just kept the old name, so I didn't really choose `ObjectStorage` with some particular intent. I'm happy to change it.

Reply:
I went with `BucketKeyStorage`. I think because it subclasses `tinydb.Storage` and it's in a module that implements a subclass of `TinyDBMetaStore`, I don't we need to include "TinyDB" in the name.

Reply:
Sounds good - can come back to if we need later on.